File: src/com/rs/game/content/controller/ControllerHandler.java
package com.rs.game.content.controller;

import com.rs.game.content.controller.impl.CorpBeastController;
import com.rs.game.content.controller.impl.activity.Wilderness;

import java.util.HashMap;

public final class ControllerHandler {
	
	/**
	 * The map of cached controllers
	 */
	private static final HashMap<Object, Class<Controller>> CACHED_CONTROLLERS = new HashMap<>();
	
	/**
	 * Reloads all controllers
	 */
	public static void reload() {
		CACHED_CONTROLLERS.clear();
		registerAll();
	}
	
	/**
	 * Registers all controllers to the map
	 */
	@SuppressWarnings("unchecked")
	public static void registerAll() {
		try {
			register("Wilderness", Wilderness.class);
			register("CorpBeastController", CorpBeastController.class);
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		}
	}
	
	@SuppressWarnings("unchecked")
	private static void register(String name, Class clazz) throws ClassNotFoundException {
		CACHED_CONTROLLERS.put(name, (Class<Controller>) Class.forName(clazz.getCanonicalName()));
	}
	
	/**
	 * Finds a controller by a key
	 *
	 * @param key
	 * 		The key
	 */
	public static Controller getController(Object key) {
		if (key instanceof Controller) {
			return (Controller) key;
		}
		Class<Controller> classC = CACHED_CONTROLLERS.get(key);
		if (classC == null) {
			return null;
		}
		try {
			return classC.newInstance();
		} catch (InstantiationException | IllegalAccessException e) {
			e.printStackTrace();
		}
		return null;
	}
}

File: src/main/java/game/content/entity/actor/player/controller/ControllerHandler.kt
package game.content.entity.actor.player.controller

import com.github.michaelbull.logging.InlineLogger
import game.content.entity.actor.player.controller.impl.CorpBeastController
import game.content.entity.actor.player.controller.impl.activity.Wilderness
import game.content.entity.actor.player.controller.impl.activity.pvp.PvPWorld
import java.util.*

object ControllerHandler {
    /**
     * The map of cached controllers
     */
    private val CACHED_CONTROLLERS = HashMap<Any, Class<Controller>>()

    /**
     * Reloads all controllers
     */
    fun reload() {
        CACHED_CONTROLLERS.clear()
        registerAll()
    }

    /**
     * Registers all controllers to the map
     */
    fun registerAll() {
        try {
            register("Wilderness", Wilderness::class.java)
            register("CorpBeastController", CorpBeastController::class.java)
            register("PvPWorld", PvPWorld::class.java)
            logger.info { "Successfully registered ${CACHED_CONTROLLERS.size} controllers." }
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }

    @Throws(ClassNotFoundException::class)
    private fun register(name: String, clazz: Class<*>) {
        CACHED_CONTROLLERS[name] =
            Class.forName(clazz.canonicalName) as Class<Controller>
    }

    /**
     * Finds a controller by a key
     *
     * @param key The key
     */
    fun getController(key: Any?): Controller? {
        if (key is Controller) {
            return key
        }
        val classC = CACHED_CONTROLLERS[key] ?: return null
        try {
            return classC.newInstance()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return null
    }

    private val logger = InlineLogger()
}